Source/WebCore/loader/FrameLoader.cpp

```diff
@@ -4742,17 +4746,40 @@ void FrameLoader::updateNavigationAPIEntries()
     RefPtr page = m_frame->page();
     if (!page)
         return;
-    if (RefPtr currentItem = m_frame->history().currentItem()) {
-        Vector<Ref<HistoryItem>> historyItems;
-        if (currentItem == page->backForward().currentItem()) {
-            for (int index = -1 * static_cast<int>(page->backForward().backCount()); index <= static_cast<int>(page->backForward().forwardCount()); index++) {
-                if (RefPtr item = page->backForward().itemAtIndex(index))
-                    historyItems.append(item.releaseNonNull());
-            }
-        } else
-            historyItems.append(*currentItem);
-        domWindow->protectedNavigation()->initializeEntries(*currentItem, historyItems);
+    RefPtr currentItem = m_frame->history().currentItem();
+    if (!currentItem)
+        return;
+
+    // https://html.spec.whatwg.org/multipage/browsing-the-web.html#getting-session-history-entries-for-the-navigation-api
+    Vector<Ref<HistoryItem>> entriesForNavigationAPI;
+    entriesForNavigationAPI.append(*currentItem);
+
+    auto rawEntries = page->backForward().allItems();
+    auto startingIndex = rawEntries.find(*currentItem);
+    if (startingIndex != notFound) {
+        Ref startingOrigin = SecurityOrigin::create(rawEntries[startingIndex]->url());
+
+        int64_t i = startingIndex - 1;
+        while (i >= 0) {
+            Ref item = rawEntries[i];
```

Comment:
I feel like this needs more bounds checks.  startingIndex can be 0, then i would be -1, and I'm not sure if we're doing the right thing here.

Reply:
To visualize what this is doing, if we navigate to `bar.org#1`:

```
[ foo.org, bar.org, bar.org#1, bar.org/2, baz.com]
                    ^ startIndex
           ^ walk back until different origin
                                ^ walk forward until different origin

```

Reply:
I believe its correct but I agree this can be cleaned up.

(if the item is the first in the list, `(startingIndex - 1 < 0) will just be false and no items are inserted before it)